### ISSUE TITLE
bootstrap config transparently switches between v2 and v3 leading to unexpected rejected configs

### ISSUE THREAD

**Issue by howardjohn** [envoyproxy/envoy#11119]
Certain changes to a bootstrap file seem to trigger Envoy to process other parts of the bootstrap differently (v2 vs v3).

Starting with this config file: 
[envoy-sds.yaml.txt](https://github.com/envoyproxy/envoy/files/4601472/envoy-sds.yaml.txt)

Validation passes

Adding a bit of config:

```yaml
dynamic_resources:
  lds_config:
    api_type: GRPC
    grpc_services:
      envoy_grpc:
        cluster_name: foo
```

Causes an unrelated part of the config to fail:
```
[2020-05-08 20:35:17.380][1][critical][main] [source/server/server.cc:98] error initializing configuration '/config/envoy-sds-lds.yaml': Protobuf message (type envoy.config.bootstrap.v3.Bootstrap reason INVALID_ARGUMENT:(static_resources.clusters[1]) hosts: Cannot find field.) has unknown fields
[2020-05-08 20:35:17.380][1][info][main] [source/server/server.cc:673] exiting
Protobuf message (type envoy.config.bootstrap.v3.Bootstrap reason INVALID_ARGUMENT:(static_resources.clusters[1]) hosts: Cannot find field.) has unknown fields
```

Full broken config file: 
[envoy-sds-lds.yaml.txt](https://github.com/envoyproxy/envoy/files/4601477/envoy-sds-lds.yaml.txt)

Configs are run with `docker run -v $HOME/kube/local:/config -p 15000:15000 envoyproxy/envoy-dev -c /config/envoy-sds-lds.yaml`

```
envoy  version: 49efb9841a58ebdc43a666f55c445911c8e4181c/1.15.0-dev/Clean/RELEASE/BoringSSL
```

**Comment by howardjohn** [comment]
Just adding

```
dynamic_resources:
  lds_config:
```

is fine so must be something in ConfigSource triggering it here. I have seen this before though, so it is not specific to this field as far as I know

**Comment by mattklein123** [comment]
I'm pretty sure this is a parsing error and the issue is the config has been upgraded to v3 before the error happens. I will defer to @htuch though.

**Comment by howardjohn** [comment]
After playing with this a bit more I realized:

```
dynamic_resources:
  lds_config:
    api_type: GRPC
```

is not actually valid config in either version. So it seems like (without any understanding of the code) an error in the config causes upgrade to v3, which then gets a different error maybe?

**Comment by htuch** [comment]
Yeah, Envoy does this transparent upgrade by default, see https://www.envoyproxy.io/docs/envoy/latest/operations/cli#cmdoption-bootstrap-version for how to set this explicitly.

Does this resolve the issue?

**Comment by howardjohn** [comment]
Yeah the flag is useful. The UX here isn't great because I am randomly presented with a new error that was previously valid because I made a mistake in the config. However if we did something like displaying the v2 and v3 errors, then for v3 config users they would always see v2 errors probably, so it wouldn't make too much sense I think. Seems like the flag is the only way to get the user intent and there probably isn't too much we can determine automatically beyond that. Thanks!
